src/librustc_borrowck/diagnostics.rs

```diff
@@ -82,10 +82,52 @@ Since `MyStruct` is a type that is not marked `Copy`, the data gets moved out
 of `x` when we set `y`. This is fundamental to Rust's ownership system: outside
 of workarounds like `Rc`, a value cannot be owned by more than one variable.
 
-If we own the type, the easiest way to address this problem is to implement
-`Copy` and `Clone` on it, as shown below. This allows `y` to copy the
-information in `x`, while leaving the original version owned by `x`. Subsequent
-changes to `x` will not be reflected when accessing `y`.
+Sometimes we don't need to move the value, we just want a reference to it. In
+the example below, we don't actually have to move `x` to `has_zero`, we can give
+it a reference to `x` with `&` instead.
+
+```
+struct MyStruct { s: u32 }
+
+fn has_zero(mys: &MyStruct) -> bool {
+    mys.s == 0
+}
+
+fn main() {
+    let x = MyStruct{ s: 5u32 };
+    let zero = has_zero(&x);
+    println!("{} == 0 is {}", x.s, zero);
+}
+```
+
+A mutable reference can be created with `&mut`.
+
+Sometimes we don't want a reference, but a duplicate. All types marked `Clone`
+can be duplicated by calling `.clone()`. Subsequent changes to a clone do not
+affect the original variable.
+
+Most types in the standard library are marked `Clone`. If we control the
+definition of a type, we can implement `Clone` on it ourselves, as shown below.
+We can then duplicate the value of `y`, and change `x` without affecting `y`.
+
+```
+#[derive(Clone)]
+struct MyStruct { s: u32 }
```

Comment:
I am trying to figure out how to tie in the copy vs clone dichotomy a bit better here. I think it's useful to explain when you would want what. I sometimes mention a three-level grouping here:

- Default: represents a unique resource
    - Example: Money
- Implements Clone: represents a resource, but one that can be copied
    - Example: Vector, String
- Implements Copy and Clone: represents a value with no ownership semantics
    - Example: integer

I am torn between keeping the presentation as you did it -- with one type (`MyStruct`) that grows impls -- versus changing to give a bit more semantics context. For example, one could have `struct MyStruct { values: Vec<u32> }` for clone and change to `struct Point { x: u32, y: u32 }` or something as an example where `Copy` would be appropriate.

Really, I'd prefer to tag in somebody like @steveklabnik or @carols10cents here -- it seems like we should work hard to ensure that this extended error message ties in well with the way that the book explains things.

Reply:
Yeah, I think giving better examples is a good idea. Clone vs Copy is explained as heap vs stack data [in the book](https://doc.rust-lang.org/book/second-edition/ch04-01-what-is-ownership.html#stack-only-data-copy). I'm not sure this is the best way to go about it though. I like your explanation better, focusing on semantics instead of low-level memory layout which may be less familiar to a newcomer.

Reply:
I also like Niko's framing as well; the book tries to do a fairly low-level explanation, but once you grasp that, the higher-level one is better, IMO.